expression/constant.go

```diff
@@ -118,9 +150,45 @@ func (c *Constant) Eval(_ chunk.Row) (types.Datum, error) {
 	return c.Value, nil
 }
 
+// UnsafeEval returns datum maybe not match const's type.
+// but it will be cast soon, so use this to reduce duplicate cast.
+func (c *Constant) UnsafeEval(_ chunk.Row) (types.Datum, error) {
+	if c.ParamMarker != nil {
+		dt := c.ParamMarker.Value()
+		if dt.IsNull() {
+			c.Value.SetNull()
+			return c.Value, nil
+		}
+		return dt, nil
+	} else if c.DeferredExpr != nil {
+		if sf, sfOK := c.DeferredExpr.(*ScalarFunction); sfOK {
+			dt, err := sf.Eval(chunk.Row{})
+			if err != nil {
+				return c.Value, err
+			}
+			if dt.IsNull() {
+				c.Value.SetNull()
+				return c.Value, nil
+			}
+			return dt, nil
+		}
+	}
+	return c.Value, nil
+}
+
 // EvalInt returns int representation of Constant.
 func (c *Constant) EvalInt(ctx sessionctx.Context, _ chunk.Row) (int64, bool, error) {
-	if c.DeferredExpr != nil {
+	if c.ParamMarker != nil {
+		dt := c.ParamMarker.Value()
+		if dt.IsNull() {
```

Comment:
It's also duplicated with the following lines when `c.DeferredExpr` is not `nil`.